frontend/metrics.go

```diff
@@ -0,0 +1,98 @@
+package main
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	chi "github.com/go-chi/chi/v5"
+
+	"github.com/Azure/ARO-HCP/internal/metrics"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type PrometheusEmitter struct {
+	gauges   map[string]*prometheus.GaugeVec
+	counters map[string]*prometheus.CounterVec
+}
+
+func NewPrometheusEmitter() *PrometheusEmitter {
+	return &PrometheusEmitter{
+		gauges:   make(map[string]*prometheus.GaugeVec),
+		counters: make(map[string]*prometheus.CounterVec),
+	}
+}
+
+func (pe *PrometheusEmitter) EmitGauge(name string, value float64, labels map[string]string) {
+	vec, exists := pe.gauges[name]
+	if !exists {
+		labelKeys := getKeys(labels)
+		vec = prometheus.NewGaugeVec(prometheus.GaugeOpts{Name: name}, labelKeys)
+		prometheus.MustRegister(vec)
+		pe.gauges[name] = vec
+	}
+	vec.With(labels).Set(value)
+}
+
+func (pe *PrometheusEmitter) EmitCounter(name string, value float64, labels map[string]string) {
+	vec, exists := pe.counters[name]
+	if !exists {
+		labelKeys := getKeys(labels)
+		vec = prometheus.NewCounterVec(prometheus.CounterOpts{Name: name}, labelKeys)
+		prometheus.MustRegister(vec)
+		pe.counters[name] = vec
+	}
+	vec.With(labels).Add(value)
+}
+
+func getKeys(m map[string]string) []string {
```

Comment:
we should be able to replace `getKeys` with `Maps.Keys`

https://pkg.go.dev/golang.org/x/exp/maps#Keys